src/fp.rs

```diff
@@ -0,0 +1,302 @@
+        carry = result.0;
+
+        result = mul64(x[1], y[1]);
+        hi = result.0;
+        lo = result.1;
+        result = add64(lo, carry, 0);
+        lo = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        hi = result.0;
+        result = add64(zz[2], lo, 0);
+        zz[2] = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        zz[3] = result.0;
+
+        // Reduction
+        let w = self.mu.wrapping_mul(zz[0]);
+        result = mul64(p[0], w);
+        hi = result.0;
+        lo = result.1;
+        result = add64(zz[0], lo, 0);
+        zz[0] = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        carry = result.0;
+
+        result = mul64(p[1], w);
+        hi = result.0;
+        lo = result.1;
+        result = add64(lo, carry, 0);
+        lo = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        hi = result.0;
+        result = add64(zz[1], lo, 0);
+        zz[1] = result.0;
+        cc = result.1;
+        result = add64(zz[2], hi, cc);
+        zz[2] = result.0;
+        cc = result.1;
+        result = add64(zz[3], 0, cc);
+        zz[3] = result.0;
+
+        let w = self.mu.wrapping_mul(zz[1]);
+        result = mul64(p[0], w);
+        hi = result.0;
+        lo = result.1;
+        result = add64(zz[1], lo, 0);
+        zz[1] = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        carry = result.0;
+
+        result = mul64(p[1], w);
+        hi = result.0;
+        lo = result.1;
+        result = add64(lo, carry, 0);
+        lo = result.0;
+        cc = result.1;
+        result = add64(hi, 0, cc);
+        hi = result.0;
+        result = add64(zz[2], lo, 0);
+        zz[2] = result.0;
+        cc = result.1;
+        result = add64(zz[3], hi, cc);
+        zz[3] = result.0;
+
+        (zz[2] as u128) | ((zz[3] as u128) << 64)
+    }
+
+    pub fn pow(&self, x: u128, exp: u128) -> u128 {
+        let mut t = self.elem(1);
+        for i in (0..128).rev() {
+            t = self.mul(t, t);
+            if (exp >> i) & 1 != 0 {
+                t = self.mul(t, x);
+            }
+        }
+        t
+    }
+
+    pub fn inv(&self, x: u128) -> u128 {
+        self.pow(x, self.p - 2)
+    }
+
+    pub fn neg(&self, x: u128) -> u128 {
+        self.sub(self.elem(0), x)
+    }
+
+    pub fn elem(&self, x: u128) -> u128 {
+        modp(self.mul(x, self.r2), self.p)
+    }
+
+    pub fn rand_elem<R: Rng + ?Sized>(&self, rng: &mut R) -> u128 {
+        let uniform = rand::distributions::Uniform::from(0..self.p);
+        self.elem(uniform.sample(rng))
+    }
+
+    pub fn from_elem(&self, x: u128) -> u128 {
+        modp(self.mul(x, 1), self.p)
+    }
+}
+
+fn lo64(x: u128) -> u64 {
+    (x & ((1 << 64) - 1)) as u64
+}
+
+fn hi64(x: u128) -> u64 {
+    (x >> 64) as u64
+}
+
+fn add64(x: u64, y: u64, carry: u64) -> (u64, u64) {
+    let sum = x.wrapping_add(y).wrapping_add(carry);
```

Comment:
it might be better to use [overflowing_add](https://doc.rust-lang.org/std/primitive.u64.html#method.overflowing_add)

Reply:
Why would this be better? Is the code incorrect as-is?

Reply:
that's the point, that someone needs to check whether this is correct, rather than using the provided function.